crmsh/sbd.py

```diff
@@ -3,14 +3,265 @@
+            shutil.rmtree(SBD_SYSTEMD_DELAY_START_DIR)
+            bootstrap.csync2_update(SBD_SYSTEMD_DELAY_START_DIR)
+            utils.cluster_run_cmd("systemctl daemon-reload")
+            return
+
+        # TimeoutStartUSec default is 1min 30s, need to parse as seconds
+        cmd = "systemctl show -p TimeoutStartUSec sbd --value"
+        out = utils.get_stdout_or_raise_error(cmd)
+        res_seconds = re.search("(\d+)s", out)
+        default_start_timeout = int(res_seconds.group(1)) if res_seconds else 0
+        res_min = re.search("(\d+)min", out)
+        default_start_timeout += 60 * int(res_min.group(1)) if res_min else 0
+        if default_start_timeout >= self.sbd_delay_start_value:
+            return
+
+        utils.mkdirp(SBD_SYSTEMD_DELAY_START_DIR)
+        sbd_delay_start_file = "{}/sbd_delay_start.conf".format(SBD_SYSTEMD_DELAY_START_DIR)
+        utils.str2file("[Service]\nTimeoutSec={}".format(int(1.2*self.sbd_delay_start_value)), sbd_delay_start_file)
+        bootstrap.csync2_update(SBD_SYSTEMD_DELAY_START_DIR)
+        utils.cluster_run_cmd("systemctl daemon-reload")
+
+    def adjust_stonith_timeout(self):
+        """
+        Adjust stonith-timeout property
+        """
+        utils.set_property(stonith_timeout=self._get_stonith_timeout())
+
+    def adjust_pcmk_delay_max(self):
+        """
+        Adjust pcmk_delay_max parameter for sbd ra
+        """
+        # TODO this function should be outside of sbd.py, to adjust any fence device
+
+        if not utils.has_resource_configured(SBDManager.SBD_RA):
+            return
+
+        if self.two_node_without_qdevice:
+            cmd = "crm resource param {} set pcmk_delay_max {}s".format(SBDManager.SBD_RA_ID, self.pcmk_delay_max)
+        else:
+            cmd = "crm resource param {} delete pcmk_delay_max".format(SBDManager.SBD_RA_ID)
+        utils.get_stdout_or_raise_error(cmd)
+
+    def adjust_delay_start(self):
+        """
+        Adjust SBD_DELAY_START in /etc/sysconfig/sbd
+        """
+        value = str(self.sbd_delay_start_value) if self.two_node_without_qdevice else "no"
+        SBDManager.update_configuration({"SBD_DELAY_START": value})
+
+    @classmethod
+    def adjust_sbd_timeout_related_cluster_configuration(cls, removing=False):
+        """
+        Adjust sbd timeout related configurations
+        """
+        if not utils.service_is_active("sbd.service") or not utils.detect_virt():
+            return
+
+        cls_inst = cls(removing=removing)
+        cls_inst._load_configurations()
+
+        message = "Adjusting sbd related timeout values for 2-node cluster"
+        if not cls_inst.two_node_without_qdevice:
+            message = "Reverting sbd related timeout values for non 2-node cluster"
+
+        with logger_utils.status_long(message):
+            cls_inst.adjust_delay_start()
+            cls_inst.adjust_pcmk_delay_max()
+            cls_inst.adjust_stonith_timeout()
```

Comment:
adjust_stonith_timeout()
adjust_pcmk_delay_max()

These two has nothing related to the virtualization environment, must run before detect_virt().

And they doesn't belong to sbd.py

Reply:
`adjust_stonith_timeout` in SBDTimeout class is for sbd, it get runtime stonith timeout value from msgwait or SBD_WATCHDOG_TIMEOUT, then compare with the value form cib property, at last use the larger one

`bootstrap.adjust_stonith_timeout_for_all_situations` will be called after above, still compare with the value from cib proerty, at last use the larger one

Reply:
Okay, detect_virt() finds it appropriate places elsewhere. 

Agree with you, adjust_stonith_timeout() do have the direct relationship with SBD.

I agree to close this conversation and defer the TODO to adjust_pcmk_delay_max() function